### PR TITLE
POC change to use c11 to implement skip list

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/include/column_inline.h [hunk 1]
```diff
@@ -19,9 +19,10 @@ __col_insert_search_gt(WT_INSERT_HEAD *ins_head, uint64_t recno)
     /*
      * Compiler may replace the following usage of the variable with another read.
      *
-     * Place a read barrier to avoid this issue.
+     * Place a compiler barrier to avoid this issue.
      */
-    WT_ORDERED_READ_WEAK_MEMORDER(ins, WT_SKIP_LAST(ins_head));
+    ins = WT_SKIP_LAST(ins_head);
+    WT_C11_BARRIER();
```
Comment (dgomezferro) [inline src/include/column_inline.h:25]: If we are worried about this, we should make `ins->tail` atomic instead of using a barrier.

File: src/include/serial_inline.h [hunk 2]
```diff
@@ -36,10 +36,12 @@ __insert_simple_func(
          * against the next pointer might indicate that the skip list location is still valid, but
          * that may no longer be true when the atomic_cas operation executes.
          *
-         * Place a read barrier here to avoid this issue.
+         * Place a compiler barrier here to avoid this issue.
          */
-        WT_ORDERED_READ(old_ins, *ins_stack[i]);
-        if (old_ins != new_ins->next[i] || !__wt_atomic_cas_ptr(ins_stack[i], old_ins, new_ins))
+        old_ins = *ins_stack[i];
+        WT_C11_BARRIER();
```
Comment (dgomezferro) [inline src/include/serial_inline.h:42]: Same thing here for `ins_stack`.

File: src/include/serial_inline.h [hunk 3]
```diff
@@ -36,10 +36,12 @@ __insert_simple_func(
...
+        if (old_ins != new_ins->next[i] ||
+          !__wt_atomic_memorder_cas_ptr(ins_stack[i], old_ins, new_ins, WT_ATOMIC_SEQ_CST))
```
Comment (RedBeard0531) [inline src/include/serial_inline.h:44]: Does this really need SEQ_CST? My best guess is that a RELEASE is sufficient, but again I don't fully understand the context.

At the very least, I'm pretty sure that you would be fine with RELAXED on the failure case, especially since you don't use the actual value discovered by the swap.

File: src/include/serial_inline.h [hunk 4]
```diff
@@ -79,10 +81,12 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
...
             return (i == 0 ? WT_RESTART : 0);
         if (ins_head->tail[i] == NULL || ins_stack[i] == &ins_head->tail[i]->next[i])
```
Comment (RedBeard0531) [inline src/include/serial_inline.h:91]: ditto all comments above. Even more so here since `ins_stack[i]` is also used after the cas. Are we intentionally forcing the compiler to reload that value? I *hope* it can't change, since I think that would have to be a data race.

**Review by RedBeard0531 — COMMENTED** [review]
Why define our own wrappers around the __atomic builtins rather than just using the c atomic API directly? 

Also, even if you don't want to use the C `_Atomic(T)` marker, it might be worth doing our own `#define WT_ATOMIC(Type) Type` so that we can start to mark arguments locally for which are intended to be atomic and which are not.

File: src/btree/bt_curprev.c [hunk 5]
```diff
@@ -87,9 +87,10 @@ __cursor_skip_prev(WT_CURSOR_BTREE *cbt)
              * code. Here we don't need to worry about CPU reordering as we are reading a thread
              * local value.
              *
-             * Place a read barrier to avoid this issue.
+             * Place a compiler barrier to avoid this issue.
              */
-            WT_ORDERED_READ_WEAK_MEMORDER(ins, cbt->ins_head->head[i]);
+            ins = cbt->ins_head->head[i];
+            WT_C11_BARRIER();
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/btree/bt_curprev.c:93]: Yeah, the comment is wrong here. cbt->ins_head->head[i] is not thread local.

File: src/include/column_inline.h (hunk 1 above)
Comment (quchenhao) [reply src/include/column_inline.h:25]: I don't really follow. line 32 is reading through a pointer we already got. Can it be reordered before we actually read the pointer `ins = WT_SKIP_LAST(ins_head)`?

Here we only intend to not read WT_SKIP_LAST(ins_head) to ins multiple times. I don't think we need to worry about the order as there is a single thread happen-before relationship?

File: src/include/column_inline.h [hunk 6]
```diff
@@ -40,11 +41,8 @@ __col_insert_search_gt(WT_INSERT_HEAD *ins_head, uint64_t recno)
         /*
          * CPUs with weak memory ordering may reorder the reads which may lead us to read a stale
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/column_inline.h:42]: Totally agree we need to change the comment to fit the C11 model as well. But I'll keep as it is for now.

File: src/include/gcc.h [hunk 7]
```diff
@@ -98,12 +98,26 @@
...
-    {                                                                   \
-        return (WT_ATOMIC_CAS(vp, &old, newv));                         \
+#define WT_ATOMIC_RELAXED __ATOMIC_RELAXED
+#define WT_ATOMIC_CONSUME __ATOMIC_CONSUME
+#define WT_ATOMIC_ACQUIRE __ATOMIC_ACQUIRE
+#define WT_ATOMIC_RELEASE __ATOMIC_RELEASE
+#define WT_ATOMIC_ACQ_REL __ATOMIC_ACQ_REL
+#define WT_ATOMIC_SEQ_CST __ATOMIC_SEQ_CST
+
+#define WT_C11_BARRIER() __atomic_signal_fence(WT_ATOMIC_ACQ_REL)
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/gcc.h:108]: deleted.

File: src/include/gcc.h [hunk 8]
```diff
@@ -125,21 +139,37 @@ WT_ATOMIC_CAS_FUNC(size, size_t *vp, size_t old, size_t newv)
 static inline bool
 __wt_atomic_cas_ptr(void *vp, void *old, void *newv)
 {
-    return (WT_ATOMIC_CAS((void **)vp, &old, newv));
+    return (WT_ATOMIC_CAS((void **)vp, &old, newv, WT_ATOMIC_SEQ_CST));
+}
+
+#define WT_ATOMIC_LOAD_PTR(type, vp, memorder) (type *)WT_ATOMIC_LOAD((void **)vp, memorder)
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/gcc.h:145]: It is really a pain if we have to do casting every time we read a pointer. Reading a pointer is going to be the predominant pattern we use for C11 in WiredTiger.

File: src/include/gcc.h [hunk 9]
```diff
@@ -125,21 +139,37 @@ WT_ATOMIC_CAS_FUNC(size, size_t *vp, size_t old, size_t newv)
...
+
+/*
+ * __wt_atomic_memorder_cas_ptr --
+ *     Pointer compare and swap with memory ordering.
+ */
+static inline bool
+__wt_atomic_memorder_cas_ptr(void *vp, void *old, void *newv, int memorder)
+{
+    return (WT_ATOMIC_CAS((void **)vp, &old, newv, memorder));
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/gcc.h:154]: I think this is because it is an implicit cast from any pointer type to void* so we declare it as void * and then cast it to void **.

File: src/include/serial_inline.h [hunk 10]
```diff
@@ -36,10 +36,12 @@ __insert_simple_func(
          * against the next pointer might indicate that the skip list location is still valid, but
          * that may no longer be true when the atomic_cas operation executes.
          *
-         * Place a read barrier here to avoid this issue.
+         * Place a compiler barrier here to avoid this issue.
          */
-        WT_ORDERED_READ(old_ins, *ins_stack[i]);
-        if (old_ins != new_ins->next[i] || !__wt_atomic_cas_ptr(ins_stack[i], old_ins, new_ins))
+        old_ins = *ins_stack[i];
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/serial_inline.h:41]: This is to prevent old_ins being read twice in __wt_atomic_memorder_cas_ptr(ins_stack[i], old_ins, new_ins, WT_ATOMIC_SEQ_CST) this call.

File: src/include/serial_inline.h (hunk 3 above)
Comment (quchenhao) [reply src/include/serial_inline.h:44]: I think we at least we need acquire release? We need to synchronise the read with another write as well?

I think we should not allow any read or write being reordered across this.

File: src/include/serial_inline.h (hunk 4 above)
Comment (quchenhao) [reply src/include/serial_inline.h:91]: ins_stack is a thread local variable so it should be ok

File: src/include/serial_inline.h [hunk 11]
```diff
@@ -79,10 +81,12 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
...
+        WT_C11_BARRIER();
+        if (old_ins != new_ins->next[i] ||
+          !__wt_atomic_memorder_cas_ptr(ins_stack[i], old_ins, new_ins, WT_ATOMIC_SEQ_CST))
             return (i == 0 ? WT_RESTART : 0);
         if (ins_head->tail[i] == NULL || ins_stack[i] == &ins_head->tail[i]->next[i])
             ins_head->tail[i] = new_ins;
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/serial_inline.h:92]: It depends on where the data is defined. If it's on session or cursor, it is thread local. If the data is on btree, it is shared.

File: src/include/column_inline.h (hunk 1 above)
Comment (RedBeard0531) [reply src/include/column_inline.h]: > line 32 is reading through a pointer we already got. Can it be reordered before we actually read the pointer ins = WT_SKIP_LAST(ins_head).

Yes. Well, more precisely, the read through the pointer isn't guaranteed to happen after the write to that location unless you have a proper chain of sequencing like the following:

1. `*producer_private = value;`
2. `public_ptr.store(producer_private, release);`
3. `consumer_private = public_ptr.load(acquire);` where that reads the value written in step 2 (or something in it release sequence)
4. `value = *consumer_private;`

What we really care about is that step 1 _happens before_ step 4. In the C memory model, that is ensured by having step 3 do a load-acquire that reads the value written by step 2. See http://eel.is/c++draft/atomics.order#2. If you want a short hard way to remember and reason about this rule, remember that acquire and release operations don't do anything on their own. They only "activate" in pairs, and only for the pair of threads involved in the pairing (or others that transitively establish an ordering with them through additional operations). 

Note that using `seq_cst` doesn't change this, it just adds a rule that in addition to the acquire-release ordering, there is a single total ordering between all `seq_cst` operations, which means that `seq_cst` is only stronger than acquire/release when paired with other `seq_cst` opertaions.

File: src/include/gcc.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (RedBeard0531) [reply src/include/gcc.h:143]: Why do you need to cast? Doesn't `WT_ATOMIC_LOAD` propagate the type information from `ptr` to the result? The `atomic_load()` function and the `__atomic` builtins are all magically generic and correctly propagate the types of their arguments to their return values, even though that isn't something you can do in a normal C function (but is very common in C++). I'm just proposing that we take advantage of that, rather than throwing away the type-safety it provides.

File: src/include/gcc.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (RedBeard0531) [reply src/include/gcc.h]: Yes, it is legal, but seems fishy. For example, it would allow this incorrect code:

```c
int* atomic;
int* old = atomic;
int* new = malloc(4); // pretend I initialized the memory
__wt_atomic_memorder_cas_ptr(atomic, old, new, release); // oops! should have used &atomic
```

File: src/include/serial_inline.h (hunk 3 above)
Comment (RedBeard0531) [reply src/include/serial_inline.h]: > I think we at least we need acquire release?  We need to synchronise the read with another write as well?

Is the write you are concerned with the next pass of the loop? If so, that would be handled by *its* release store happening after everything that occurred earlier on the same thread (including this release store). Is there a different write you are concerned with?

Also, I think we already have an acquire edge for this location/value pair if I understand this code correctly. At some point earlier on this thread did we do a load acquire on that address to get the value that we wrote into `new_ins->next[i]`? If so, then this CAS is already guaranteed to happen after the write of that value. And because of the CAS we know that the value must not have changed since then, so we don't need to re-establish the acquire edge to that address. I am assuming that we don't have to worry about the ABA problem here because we prevent freeing theses addresses while we are in there (I think that is handled by hazard pointers to the whole page).

> I think we should not allow any read or write being reordered across this.

Remember that the C model doesn't allow you to express exactly that. You need to think in terms of paired acquires and releases.

File: src/include/serial_inline.h (hunk 4 above)
Comment (RedBeard0531) [reply src/include/serial_inline.h]: Sure but do we need to force a reload, or should it be allowed to store `ins_stack[i]` in a register? I assume we want to allow that, so we should consider assigning it to a variable like suggested in https://github.com/wiredtiger/wiredtiger/pull/9010#discussion_r1157157035

File: src/include/serial_inline.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (RedBeard0531) [reply src/include/serial_inline.h:92]: It is a bit confusing because you can't see what things are on the cursor or session locally in this function. Also, if I understand correctly (not sure that I do!) `ins_stack[i]` itself is stored on the cursor, but it is a pointer to something stored in the btree?

Based on the name, I had assumed that `ins_head` was a pointer to the head of the skiplist which I assumed was stored in the tree. Is that correct? If so, doesn't that mean that this needs to be an atomic access and probably needs a memory ordering?

File: src/include/serial_inline.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (RedBeard0531) [reply src/include/serial_inline.h]: To be precise, you don't care how many times `old_ins` (the local variable) is loaded, but you *do* care that is has the same value on both sides of the `||` condition. Doing `atomic_load(ins_stack[i], relaxed)` is sufficient to ensure that. I don't know for sure if relaxed here is enough for the algorithm, but based on other comments, I suspect it is.

File: src/include/column_inline.h (hunk 1 above)
Comment (RedBeard0531) [reply src/include/column_inline.h]: PS - what I think you are describing is something that hardware often guarantees (with the notable exception of the DEC Alpha) under the name "address dependency ordering".  It basically says that any reads that contribute to the value of an address must happen before a read of that address. Arm also provides this guarantee at the assembly level, however it has weird limitations around transitivity relative to using acquire loads as can be seen [here](https://wg21.link/p0735r1).

There was an attempt to provide access to this guarantee in the C memory model as `memory_order_consume`, however compiler implementors have found it basically impossible to preserve the guarantees from source code all the way to generated assembly so they all just implement it with the same codegen as acquire, which is why I say it is usually best to pretend that consume doesn't exist. There have been several attempts at fixing consume for c++ (see [P0750](https://wg21.link/p0750) along with the links in its background section), but AFAICT there hasn't been much movement on that front recently). The LKMM (linux kernel memory model) tries to allow developers to take advantage of this hardware guarantee, but you can see all of the caveats that they document [here](https://github.com/torvalds/linux/blob/master/Documentation/RCU/rcu_dereference.rst). I do not recommend we attempt to duplicate this, unless we find some *really* hot code path where even `LDAPR` is too expensive.

File: src/include/serial_inline.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/serial_inline.h:92]: bq. Based on the name, I had assumed that ins_head was a pointer to the head of the skiplist which I assumed was stored in the tree. Is that correct? If so, doesn't that mean that this needs to be an atomic access and probably needs a memory ordering?

In each element in the skip list, there is an array next storing the pointers to the next element at each level in the skip list. ins_head stores the addresses of that array element.

In other words, it stores the pointer to the pointer of a skip list element. The address itself doesn't change even there is concurrent update to the skip list.

File: src/include/gcc.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/gcc.h:143]: It's a surprise to me that just using `WT_ATOMIC_LOAD` actually works. Thanks for the tip.

File: src/include/column_inline.h (hunk 1 above)
Comment (quchenhao) [reply src/include/column_inline.h]: Then I think in addition to an acquire read here, we also need a release store when we write to the head.

File: src/include/serial_inline.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/serial_inline.h]: I agree. It is safer to use an acquire read here.

File: src/include/serial_inline.h (hunk 3 above)
Comment (quchenhao) [reply src/include/serial_inline.h]: I think you are right in the new c11 model context. However, the old wt memory model requires cas to give a full barrier semantic. This is why I'm a bit hesitant to relax the order here. As experiment, I'll have a try.

File: src/include/serial_inline.h (hunk 4 above)
Comment (quchenhao) [reply src/include/serial_inline.h]: agree

File: src/include/btmem.h [hunk 12]
```diff
@@ -1447,11 +1447,22 @@ struct __wt_insert {
  */
 #define WT_SKIP_FIRST(ins_head) \
     (((ins_head) == NULL) ? NULL : ((WT_INSERT_HEAD *)(ins_head))->head[0])
+#define WT_SKIP_FIRST_ATOMIC(ins_head) \
```
Comment (agorrod) [inline src/include/btmem.h:1450]: I was hoping that we'd end up with quite a different approach to this change. Rather than swapping out primitives where we already have them, to change variables so that they are marked as atomic and the compiler (and/or tooling) encourage us to use them safely.
I understand that there are some places where skip lists don't currently require coordination because they are protected by locks, but as far as I can tell that was probably done for simplicity rather than performance.

Comment (quchenhao) [reply src/include/btmem.h:1450]: In addition to this change, I'm also trying to declare the variable as atomic but that is harder than I thought due to compiler issues. However, from my understanding, I don't think the compiler or tooling will do anything magic even we do that. They will just use the strictest order by default where the variable is used. In this case, we need to explicitly tell the compiler to use the appropriate order in places it is too strict.

This is a poc change, I'm introducing `WT_SKIP_FIRST_ATOMIC` just to see what is the minimum change required. I think in the new model, we would have macros like `WT_SKIP_FIRST_RELAXED` and `WT_SKIP_FIRST_ACQUIRE` instead.

Comment (agorrod) [reply src/include/btmem.h:1450]: > I don't think the compiler or tooling will do anything magic even we do that

That makes me a bit sad. I'd hoped if we declared a variable as C11 atomic we'd be forced to use safe semantics when accessing/updating it.

> I'm introducing WT_SKIP_FIRST_ATOMIC just to see what is the minimum change required. I think in the new model, we would have macros like WT_SKIP_FIRST_RELAXED and WT_SKIP_FIRST_ACQUIRE instead.

Interesting. It's not really my preferred path - since it isn't obvious when I'd choose between relaxed and acquire from their names. i.e: I need an in depth understanding of the semantics to know how to choose which one. It could just be a naming game, but I'd really like for this to be simpler (less ways to access/update a shared variable) rather than more.

Comment (RedBeard0531) [reply src/include/btmem.h:1450]: >> I don't think the compiler or tooling will do anything magic even we do that
> 
> That makes me a bit sad. I'd hoped if we declared a variable as C11 atomic we'd be forced to use safe semantics when accessing/updating it.

Note that `_Atomic(T)` variables do fail _safe_, so that if you don't explicitly use any atomic access (eg using `atomic_var = value` or `value = atomic_var`, or even `atomic_var += value`), they behave equivalently to using `seq_cst`. If you don't want this implicit atomic access, you can do something like this to force correct explict accesses of marked variable:

```c
#define WT_DEFINE_ATOMIC_TYPE(name, T) \
  typedef struct { \
      _Atomic(T) atomic_var; \
  } name

// There are some preprocessor tricks if you change `order` to `...` to make it optional.
// Not sure if that is what you would want or not.
#define WT_ATOMIC_LOAD(var, order) atomic_load_explicit((var).atomic_var, memory_order_##order)
#define WT_ATOMIC_STORE(var, value, order) atomic_load_explicit((var).atomic_var, (value), memory_order_##order)
// And all the other atomic_blah functions

WT_DEFINE_ATOMIC_TYPE(WT_ATOMIC_INT, int);
WT_DEFINE_ATOMIC_TYPE(WT_ATOMIC_PTR_INSERT, WT_INSERT*);


WT_ATOMIC_PTR_INSERT inserts; // in some structure somewhere

WT_INSERT* local_inserts; // a local var
if ((local_inserts = WT_ATOMIC_LOAD(inserts, acquire) != NULL) {
  use(local_inserts);
}
```

The advantage of this style is that you can always see which accesses are potentially concurrent when reading the code (assuming of course that every variable is correctly marked!). Having spent time in both a codebase that required using explicit loads and stores (mongodb) and one that often relies on the accesses of atomics implicitly being `seq_cst`, my personal experience is that it is much easier to reason about code using the explicit style. This is just about whether the accesses are "marked"; mongo still uses `.load()` and `.store(val)` methods that imply `seq_cst` rather than requiring explict memory orders, but having them as visual markers that some access is potentially concurrent and not subject to normal refactorings that you can do in single threaded code, is really nice in my experience.

Of course all of this (both the compiler-enforced explicit access and the implicitly-safe access through an `_Atomic(T)`) goes out the window if you have widespread casting of pointers in your codebase. This applies to both explict casting like `*(int*)(some_atomic_int_pointer)` and the implicit casting that you get if you ever end up with a `void*` in C, like `void* vp = &some_atomic_int; int* ip = vp; *ip = 7; // Oh no!`. The middle line is silently allowed in C, but thankfully not in C++.

>> I'm introducing WT_SKIP_FIRST_ATOMIC just to see what is the minimum change required. I think in the new model, we would have macros like WT_SKIP_FIRST_RELAXED and WT_SKIP_FIRST_ACQUIRE instead.
>
> Interesting. It's not really my preferred path - since it isn't obvious when I'd choose between relaxed and acquire from their names. i.e.: I need an in depth understanding of the semantics to know how to choose which one. It could just be a naming game, but I'd really like for this to be simpler (less ways to access/update a shared variable) rather than more.

I don't really love that either, but for a different reason. I don't think we can really get around needing to know about `acquire` vs `release` vs `seq_cst` vs `relaxed`. We _could_ just use `seq_cst` everywhere, but since it is defined in terms of `acquire`/`release` plus adding a total ordering over all `seq_cst` ops, I think you still need to understand the fundamental concepts any time you are doing much more than simple counters with atomics. We could also rename them, but I suspect that will cause much more confusion than it would save.

My objection to that style is that it is completely non-orthogonal and requires a combinatorial explosion of those #defines. You need them for each atomic member times each access style (load, store, cmpxchg, etc), times each memory order. I think it makes a lot more sense to keep those things orthogonal and let the expression determine the member/variable, let the function/macro determine the access style, and an optional explicit memory order argument determine the ordering. Of course, that could be my C++-trained brain talking. I know in C, non-generic, non-orthogonal code is much more common. And it seems like we already define "accessor macros" in several places like https://github.com/wiredtiger/wiredtiger/blob/b7d303e85390633f5d1c8cbf6296c370c0b5115a/src/include/btmem.h#L1430-L1440

(Those specific ones really scare me because they embed/hide a forced `(WT_INSERT*)` cast, so they can hide typing bugs from the compiler. But it seems like that is already widely accepted in the codebase, so I don't know if now is a good time to move away from that style. And it could also be me trying to use C++ sensibilities on a C codebase...)

File: src/include/serial_inline.h (hunk 3 above)
Comment (RedBeard0531) [reply src/include/serial_inline.h]: > However, the old wt memory model requires cas to give a full barrier semantic.

So there are two forms of "requirements" and I think it is important to separate them:
1. The memory model has a definition that says that nothing at all can be reordered across a CAS. This is a requirement placed on the platform-specific implementations of the atomic primitives. And it is something that all algorithms can rely on whether they need it or not.
2. Each algorithm has its own specific requirements of what it needs from each operation. When translating the abstract algorithm to code in a specific memory model, you need to choose which operations to use to ensure that those operations meet the requirements of the algorithm. Of course, that is simpler if all ops have full barrier, totally-ordered semantics.

I think you are using "requires" in the sense of 1 here. But since the whole point is that we are transitioning between memory models, I think we really need to be thinking about 2. Rather than saying that the existing code has a full barrier and we must match it, I think it would be better to figure out what the algorithm really requires, and use those specific orderings. Of course it is possible that `seq_cst` really _is_ what the algorithm needs, but I think in many more cases we are getting it accidentally because it was the only option available in the memory model. And we can still choose to use `seq_cst` in most places out of an abundance of caution, but I think the work to reason through what the least-restrictive ordering is for all accesses (possibly excluding debug-only checks) will pay off.

File: src/include/serial_inline.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (RedBeard0531) [reply src/include/serial_inline.h]: >  The address itself doesn't change even there is concurrent update to the skip list.

Sure. But I'm specifically referring to the pointer stored at `ins_head->tail[i]`, not just `ins_head` itself. I assume that the `i`th `tail` pointer needs to be accessed with at least store-release and load-acquire semantics.

File: src/include/column_inline.h (hunk 1 above)
Comment (RedBeard0531) [reply src/include/column_inline.h]: > Then I think in addition to an acquire read here, we also need a release store when we write to the head.

Exactly! Do we not have that right now?

Comment (quchenhao) [reply src/include/column_inline.h]: We have that for the insert list head as it is changed by compare and swap but we miss that in the tail. I fixed that in the pr.

File: src/include/gcc.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (RedBeard0531) [reply src/include/gcc.h]: https://en.cppreference.com/w/c/atomic/atomic_load
> This is a [generic function](https://en.cppreference.com/w/c/language/generic) defined for all [atomic object types](https://en.cppreference.com/w/c/language/atomic) A.

The [generic function](https://en.cppreference.com/w/c/language/generic) link shows how we can do something similar in macros we define, although it requires some trickery to be _as_ generic as those functions without being built into the compiler. I'm not 100% sure it is possible in plain C.

File: src/include/serial_inline.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (RedBeard0531) [reply src/include/serial_inline.h]: Looks like this was already addressed by https://github.com/wiredtiger/wiredtiger/pull/9010/commits/1ae6a0bc084fa9e65ae8df99605e6b5ea862498a#diff-302b65ca76325e7e93eb73d95d1993931e1b90d3a660d8a5fd4417c7d005ae3eR91-R93

File: src/include/column_inline.h (hunk 1 above)
Comment (RedBeard0531) [reply src/include/column_inline.h]: > We have that for the insert list head as it is changed by compare and swap but we miss that in the tail. I fixed that in the pr.

Are you referring to [this](https://github.com/wiredtiger/wiredtiger/pull/9010/commits/1ae6a0bc084fa9e65ae8df99605e6b5ea862498a#diff-302b65ca76325e7e93eb73d95d1993931e1b90d3a660d8a5fd4417c7d005ae3eR122-R127) change? If so, I think that looks like a bug even under the "old" WT memory model, and should be fixed in master, separately from this PR.

PS- finding these sorts of things is why I think that thinking in terms of finding the paired acquires and releases helps to avoid introducing memory ordering bugs, rather than just thinking about the barriers needed for each side in isolation.

Comment (quchenhao) [reply src/include/column_inline.h]: That's correct. However, I'm not sure that is actually a problem in the original wt code as anyway we never use an acquire read to read the head or tail. I think the assumption is that it is ok to read an inconsistent value from head or tail as they are the places we start walking the list and missing a few new items is ok. I changed this in the pr to make the usage of the C11 model complete.

Comment (RedBeard0531) [reply src/include/column_inline.h]: We _may_ also be saved on master by the fact that the immediately prior write was a CAS (implying a full barrier on master), so this store couldn't be reordered any earlier, even without a separate store barrier in front. But I find these sorts of "accidentally correct" cases a bit hard to reason about.

File: src/include/serial_inline.h [hunk 13]
```diff
@@ -79,13 +81,16 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
...
+        old_ins = WT_ATOMIC_LOAD(ins_stack[i], WT_ATOMIC_ACQUIRE);
+        if (old_ins != new_ins->next[i] ||
+          !__wt_atomic_memorder_cas_ptr(
+            ins_stack[i], old_ins, new_ins, WT_ATOMIC_RELEASE, WT_ATOMIC_RELAXED))
             return (i == 0 ? WT_RESTART : 0);
-        if (ins_head->tail[i] == NULL || ins_stack[i] == &ins_head->tail[i]->next[i])
-            ins_head->tail[i] = new_ins;
+        tail = WT_ATOMIC_LOAD(&ins_head->tail[i], WT_ATOMIC_ACQUIRE);
+        if (tail == NULL || ins_stack[i] == &tail->next[i])
+            WT_ATOMIC_STORE(&ins_head->tail[i], new_ins, WT_ATOMIC_RELEASE);
```
Comment (RedBeard0531) [inline src/include/serial_inline.h:93]: Staring at these lines a bit more, should this store be a CAS instead? What should happen if we raced with someone else storing a different pointer to `ins_head->tail[i]`? Is that impossible for some reason? Is it harmless? Is it a bug?
